### PR TITLE
Replace tape with tap

### LINKED ISSUE #1123: Stop using tape for tests
Tape runs all tests in the same process.  There doesn't seem to be a way to request process-per-test isolation.

Monkey-patching in tests ends up affecting other tests.  I'm quite sure at least one or two bugs (hopefully only in tests) went unnoticed because of it so I propose we move away from tape.

### PR DESCRIPTION
**Pull request by NatalieWolfe**
The [tap](https://www.npmjs.com/package/tap) module is a more fully featured TAP framework that executes its tests in separate child processes.

Resolves #1123

### CODE REVIEW

**Comment by bnoordhuis** [comment]
Thanks, this is great!  I wonder though, does `tap.test()` actually run tests in a separate process?  I glean from node-tap's source that only `tap.spawn()` does that.

**Comment by NatalieWolfe** [comment]
Correct, `tap.test` doesn't run each in their own process, but `$ tap test/test-*` will spawn a separate child process for each file that matched. Was it per-individual-test that needed process isolation?

**Comment by bnoordhuis** [comment]
Yes.

**Comment by NatalieWolfe** [comment]
Is it possible to move the tests which need isolation into individual files? Which tests need this? I can look into doing that step as part of this ticket.

**Comment by bnoordhuis** [comment]
I'm going to say 'all of them' - it's not really true but isolation never hurts.  It's my original plan of attack for #1123.

**Comment by NatalieWolfe** [comment]
Is that necessary up front? That seems like the kind of thing that could be resolved on a case-by-case basis when colliding tests are discovered. Moving individual tests into separate files isn't that hard, but seems like overkill if you don't actually need them isolated.

**Comment by bnoordhuis** [comment]
Such things tend to go unnoticed, though.  I raised #1123 after I found two unrelated tests had been clobbering each other undetected for some time.

Another anecdote: a project I was involved in had a bug caused by lack of isolation that stopped half the tests from running but it went undiscovered for a year because the CI was green all the time.

File: package.json [hunk 1]
```diff
@@ -40,12 +40,12 @@
...
-    "tape": "~4.2.0",
     "bindings": "~1.2.1",
     "nan": "^2.0.0",
-    "require-inject": "~1.3.0"
+    "require-inject": "~1.3.0",
+    "tap": "^10.3.2"
   },
   "scripts": {
-    "test": "tape test/test-*"
+    "test": "tap test/test-*"
```
Comment (refack) [inline package.json]: IMHO need to add `-Rtap`

**Comment by refack** [comment]
@NatalieWolfe do you need help? I can help...

**Comment by maclover7** [comment]
ping @NatalieWolfe

**Comment by rvagg** [comment]
needs to be updated, either @NatalieWolfe needs to revisit or someone else should open a PR, I'd +1 on merging this if it was ready

**Comment by cclauss** [comment]
@NatalieWolfe Is this already done elsewhere or is there more to do here?

**Comment by richardlau** [comment]
This was done in https://github.com/nodejs/node-gyp/pull/1795.

**Pull request closed without merging**